src/arcsf/eval/quantitative_eval.py

```diff
@@ -0,0 +1,126 @@
+import argparse
+import os
+
+import numpy as np
+import torch
+import yaml
+from torch.utils.data import DataLoader, Dataset
+from tqdm import tqdm
+from transformers import GPT2LMHeadModel, GPT2Tokenizer
+
+from arcsf.data.data_module import EvalQADataset, get_data, qa_formatter_autoregression
+from arcsf.eval.metrics import truth_ratio
+from arcsf.eval.utils import get_loss
+from arcsf.utils import get_device
+
+
+def quantitative_eval(
+    model: torch.nn.Module,
+    dataset: Dataset,
+    batch_size: int,
+    device: torch.device,
+) -> tuple[torch.Tensor, torch.Tensor]:
+    """Performs quantitative evaluation of the selected model over selected data.
+    Returns the input-wise losses on target answers for the ground truth and perturbed
+    answers.
+
+    Args:
+        model : Transformers model used to perform evaluation on
+        dataset : Dataset to perform evaluation on
+        batch_size : batch size for dataloader
+        device : Pytorch device on which to perform computation
+
+    Returns:
+        tr : truth_ratio metric from the TOFU paper
+        all_losses : input-wise losses for each ground truth and perturbed input
+    """
+    # move model to device create dataloader, initialise all_losses tensor
+    model = model.to(device)
+    n_perturbed = dataset.n_perturbed
+    data_loader = DataLoader(dataset, batch_size=batch_size, shuffle=False)
+    all_losses = torch.zeros((dataset.__len__(), n_perturbed + 1))
+    # loop over batches
+    for batch_idx, batch in enumerate(tqdm(data_loader, desc="Batch")):
+        gt_batch = batch[0]
+        batch_start_index = batch_idx * batch_size
+        batch_end_index = batch_idx * batch_size + len(gt_batch["input_ids"])
+        # don't need gradient
+        with torch.no_grad():
+            gt_dummy_model_output = model(**gt_batch)
+        # get ground truth loss
+        gt_loss = get_loss(gt_dummy_model_output.logits, gt_batch["labels"])
+        all_losses[batch_start_index:batch_end_index, 0] = gt_loss
+        # loop over perturbed samples to get their losses
+        for perturbed_index in range(1, n_perturbed + 1):
+            p_batch = batch[perturbed_index]
+            with torch.no_grad():
+                p_dummy_model_output = model(**p_batch)
+            all_losses[batch_start_index:batch_end_index, perturbed_index] = get_loss(
+                p_dummy_model_output.logits, p_batch["labels"]
+            )
+    # calculate truth_ratio and return them along with losses
+    means = torch.mean(all_losses, dim=0)
+    print(f"Mean losses:\n\tground_truth:{means[0]}\n\tperturbed:{means[1:]}")
+    tr = truth_ratio(all_losses)
+
+    return tr, all_losses
```

Comment:
Can update this to also return the probability, the `all_losses` object should allow us to calculate this elsewhere, but will implement here too. I will also investigate where Rouge score should be returned.

Reply:
Revisiting the paper, the probability is used when there are multiple choice answers. That is, the real author/world fact questions. So here we're not using them. Though, perhaps this is something that needs addressing?

Reply:
Rouge score will be better placed in qualitative eval, since the inputs are strings. Though maybe it would be better to get all of the metrics from a single script? I have issues here though, formatting the inputs when using batches.